File: vuln-sandbox/python/cwe-78/command-injection-safe.py
# SAFE: command-injection — subprocess called with a list (no shell=True) and allowlist
# Rule: InjectPythonSubprocessShell | CWE-78 | Expected: TrueNegative

import subprocess
from flask import Flask, request, jsonify

app = Flask(__name__)

ALLOWED_COMMANDS = {'ping', 'date', 'uptime'}


@app.route('/run', methods=['GET'])
def run_command():
    """Run a whitelisted command without shell injection risk."""
    cmd = request.args.get('cmd', '')

    # SAFE: command validated against allowlist; shell=False (default) prevents metacharacter injection
    if cmd not in ALLOWED_COMMANDS:
        return jsonify({'error': 'Command not allowed'}), 400

    result = subprocess.run([cmd], capture_output=True, text=True, timeout=5)
    return jsonify({'output': result.stdout, 'returncode': result.returncode})


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/rust/cwe-78/command_injection_safe.rs
// SAFE: rust-command-shell-sh — allowlist validation prevents command injection
// Rule: rust-command-shell-sh | CWE-78 | Expected: TrueNegative

use std::collections::HashSet;
use std::process::Command;

fn run_allowed_command(user_input: &str) -> Result<String, String> {
    let allowed: HashSet<&str> = ["date", "uptime", "pwd"].iter().cloned().collect();

    // SAFE: command validated against an allowlist; Command::new called directly, no sh -c
    if !allowed.contains(user_input) {
        return Err(format!("Command '{}' is not allowed", user_input));
    }

    let output = Command::new(user_input)
        .output()
        .map_err(|e| e.to_string())?;

    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

fn main() {
    let user_input = std::env::args().nth(1).unwrap_or_else(|| "date".to_string());
    match run_allowed_command(&user_input) {
        Ok(result) => println!("{}", result),
        Err(e) => eprintln!("Error: {}", e),
    }
}
